dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/DefaultFuture.java

```diff
@@ -108,11 +129,34 @@ public static void sent(Channel channel, Request request) {
         }
     }
 
+    /**
+     * close a channel when a channel is inactive
+     * directly return the unfinished requests.
+     *
+     * @param channel
+     */
+    public static void closeChannel(Channel channel) {
```

Comment:
I guess we may possibly avoid introducing `UNFINISHED_REQUESTS` by doing this:

```java
        for (long id : CHANNELS.keySet()) {
            if (channel.equals(CHANNELS.get(id))) {
                DefaultFuture future = getFuture(id);
                if (!future.isDone()) {
                    Response disconnectResponse = new Response(r.getId());
                    disconnectResponse.setStatus(Response.CHANNEL_INACTIVE);
                    disconnectResponse.setErrorMessage("Channel " + channel + " is inactive. Directly return the unFinished request.");
                    DefaultFuture.received(channel, disconnectResponse);
                }
            }
        }
```

Reply:
Looks good to me! I’ll fix it soon.